compiler/optimizer/LoopCanonicalizer.cpp

```diff
@@ -1999,6 +2068,32 @@ void TR_LoopCanonicalizer::eliminateRedundantInductionVariablesFromLoop(TR_Regio
                      }
                   }
                }
+            static bool disableTrivialArrayOperationCheck = (NULL != feGetEnv("TR_DisableTrivialArrayOperationCheck"));
+            if (!disableTrivialArrayOperationCheck)
+               {
+               TR::Node *derivedInductionVarNode = NULL;
+               ListIterator<TR::Block> blockIt(&blocksInNaturalLoop);
+               for (TR::Block *derivedInductionVarBlock = blockIt.getCurrent(); derivedInductionVarBlock; derivedInductionVarBlock = blockIt.getNext())
+                  {
+                  for (TR::TreeTopIterator treeTopIt(derivedInductionVarBlock->getEntry(),comp()); treeTopIt != derivedInductionVarBlock->getExit(); treeTopIt.stepForward())
+                     {
+                     if (treeTopIt.currentNode()->getSymbol() == derivedInductionVar->getSymbol())
+                        {
+                        derivedInductionVarNode = treeTopIt.currentNode();
+                        goto trivialArrayOperationCheck;
+                        }
+                     }
+                  }
+
+               TR_ASSERT_FATAL_WITH_NODE(derivedInductionVarNode, derivedInductionVarNode != NULL, "Unable to find node containing the derived induction variable");
+               trivialArrayOperationCheck:
+               TR::NodeChecklist visited(comp());
+               if (isTrivialArrayOperation(visited, derivedInductionVarNode))
+                  {
+                  performTransformation(comp(), "%sSkipping induction variable elimination for %d as it is being used in a trivial array operation.\n",OPT_DETAILS, derivedInductionVar->getReferenceNumber());
+                  goto moveToNextIndVar;
```

Comment:
I'm not a huge fan of this long-range goto - I'm not quite sure how to change the code structure to avoid it, but it is not very nice for reading and understanding the code.

Reply:
I could use the result from `isTrivialArrayOperation(..)` to set `incrementsInLockStep` https://github.com/eclipse/omr/blob/f4ab45dc92a334b143d5c434f441a3c950e935a6/compiler/optimizer/LoopCanonicalizer.cpp#L2109
if it is a trivial array operation `incrementsInLockStep` would be set to false and we would skip the next two if blocks